### PR TITLE
Make remote directories before put_files

### PR DESCRIPTION
**Pull request by martindurant**
(async implementations, see https://github.com/dask/s3fs/issues/491 )

### CODE REVIEW

File: fsspec/asyn.py [hunk 1]
```diff
@@ -369,12 +369,16 @@ async def _put(self, lpath, rpath, recursive=False, **kwargs):
             lpath = make_path_posix(lpath)
         fs = LocalFileSystem()
         lpaths = fs.expand_path(lpath, recursive=recursive)
-        rpaths = other_paths(lpaths, rpath)
+        dirs = [l for l in lpaths if os.path.isdir(l)]
+        rdirs = other_paths(dirs, rpath)
+        await _throttled_gather([self._makedirs(d, exist_ok=True) for d in rdirs])
```
Comment (isidentical) [inline fsspec/asyn.py:374]: Since `makedirs()` doesn't do anything relevant with fds, I think it is safe to just use `asyncio.gather` for this.

Comment (martindurant) [reply fsspec/asyn.py:374]: Is there any cost to always using _throttled (or whatever we finally call it) ?

Comment (isidentical) [reply fsspec/asyn.py:374]: We will add callback management to it, so it is better just avoid using it if not necessary (and in this case, it isnt).

File: fsspec/asyn.py [hunk 2]
```diff
@@ -369,12 +369,16 @@ async def _put(self, lpath, rpath, recursive=False, **kwargs):
             lpath = make_path_posix(lpath)
         fs = LocalFileSystem()
         lpaths = fs.expand_path(lpath, recursive=recursive)
-        rpaths = other_paths(lpaths, rpath)
+        dirs = [l for l in lpaths if os.path.isdir(l)]
+        rdirs = other_paths(dirs, rpath)
+        await asyncio.gather([self._makedirs(d, exist_ok=True) for d in rdirs])
```
Comment (isidentical) [inline fsspec/asyn.py:374]:
```suggestion
        await asyncio.gather(*[self._makedirs(d, exist_ok=True) for d in rdirs])
```

Comment (martindurant) [reply fsspec/asyn.py]: Already got it :)

**Comment by isidentical** [comment]
This change broke adlfs, since AzureBlobFileSystem doesn't implement _makedirs.

**Pull request merged**